src/compiler/scala/tools/nsc/transform/Fields.scala

```diff
@@ -513,51 +513,60 @@ abstract class Fields extends InfoTransform with ast.TreeDSL with TypingTransfor
-      val name  = lazyVal.name.toTermName.append(nme.LAZY_LOCAL_SUFFIX_STRING)
-      val holderSym =
-        lazyVal.owner.newValue(name, lazyVal.pos, flags) setInfo refTpe
-
-      val accessor = mkAccessor(lazyVal) {
-        import CODE._
-        val initializedGetter = refTpe.member(nme.initialized)
-        val setInitialized    = Apply(Select(Ident(holderSym), initializedGetter.setterIn(refClass)), TRUE :: Nil)
-        val isUnit = refClass == LazyUnitClass
-        val valueGetter = if (isUnit) NoSymbol else refTpe.member(nme.value)
-        val valueSetter = if (isUnit) NoSymbol else valueGetter.setterIn(refClass)
-        val setValue    = if (isUnit) rhs      else Apply(Select(Ident(holderSym), valueSetter), rhs :: Nil)
-        val getValue    = if (isUnit) UNIT     else Apply(Select(Ident(holderSym), valueGetter), Nil)
-
-        // must read the value within the synchronized block since it's not volatile
-        // (there's no happens-before relation with the read of the volatile initialized field)
-        // TODO: double-checked locking (https://github.come/scala/scala-dev/issues/204)
-        gen.mkSynchronized(Ident(holderSym))(
-          If(Ident(holderSym) DOT initializedGetter, getValue, Block(List(setValue, setInitialized), getValue)))
+      val refClass    = lazyHolders.getOrElse(lazyValType.typeSymbol, LazyRefClass)
+      val refTpe      = if (refClass != LazyRefClass) refClass.tpe else appliedType(refClass.typeConstructor, List(lazyValType))
+
+      val lazyName  = lazyVal.name.toTermName
+      val pos       = lazyVal.pos.focus
+      val flags     = (lazyVal.flags & FieldFlags | ARTIFACT | MUTABLE) & ~(IMPLICIT | STABLE) // TODO: why include MUTABLE???
+      val holderSym = owner.newValue(lazyName append nme.LAZY_LOCAL_SUFFIX_STRING, pos, flags) setInfo refTpe
+
+      val initializedGetter = refTpe.member(nme.initialized)
+      val isUnit            = refClass == LazyUnitClass
+      val valueGetter       = if (isUnit) NoSymbol else refTpe.member(nme.value)
+      val getValue          = if (isUnit) UNIT     else Apply(Select(Ident(holderSym), valueGetter), Nil)
+
+      def mkChecked(res: Tree) = If(Ident(holderSym) DOT initializedGetter, getValue, res)
+
+      val computerSym =
+        owner.newMethod(lazyName append nme.LAZY_SLOW_SUFFIX, pos, ARTIFACT | PRIVATE) setInfo MethodType(Nil, lazyValType)
+
+      val rhsAtComputer = rhs.changeOwner(lazyVal -> computerSym)
+      val computer = mkAccessor(computerSym) {
+        val setInitialized = Apply(Select(Ident(holderSym), initializedGetter.setterIn(refClass)), TRUE :: Nil)
+        val setValue =
+          if (isUnit) rhsAtComputer
+          else Apply(Select(Ident(holderSym), valueGetter.setterIn(refClass)), rhsAtComputer :: Nil)
+
+        gen.mkSynchronized(Ident(holderSym))(mkChecked(Block(List(setValue, setInitialized), getValue)))
```

Comment:
The issue with having `getValue` within the synchronized block is that erausre creates a box/unbox operations (because the synthetic method `def synchronized[T](body: T): T` is not specialized).

Would it be safe to just read the value after the `synchronized` block?

```
    private final /* synthetic */ int v$lzycompute$1(LazyInt v$lzy$1) {
        LazyInt lazyInt = v$lzy$1;
        synchronized (lazyInt) {
            if (!v$lzy$1.initialized()) {
                v$lzy$1.value_$eq(this.getInt());
                v$lzy$1.initialized_$eq(true);
            }
        }
        return v$lzy$1.value();
    }
```